File: Algorithms/longest-substring-without-repeating-characters.cpp
// 3. Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

/*
Given a string, find the length of the longest substring without repeating characters.

Examples:

Given "abcabcbb", the answer is "abc", which the length is 3.

Given "bbbbb", the answer is "b", with the length of 1.

Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        array<int, 512> A;
        A.fill(0);
        int sz = s.size(), i = 0, j = 0, cnt = 0, result = 0;
        while (j < sz) {
            int id = s[j];
            if (A[id] > 0) {
                ++cnt;
            }
            ++A[id];
            ++j;
            while (cnt > 0) {
                int id = s[i];
                --A[id];
                if (A[id] > 0) {
                    --cnt;
                }
                ++i;
            }
            if (result < j - i) {
                result = j - i;
            }
        }
        return result;
    }
};

File: Algorithms/longest-substring-without-repeating-characters.cs
using System;

public class Solution {
    public static int LengthOfLongestSubstring(string s) {
        int[] A = new int[512];
        int sz = s.Length, i = 0, j = 0, cnt = 0, result = 0;
        for (i = 0; i < 512; ++i) {
            A[i] = 0;
        }
        i = 0;
        while (j < sz) {
            if (A[s[j]] > 0) {
                ++cnt;
            }
            ++A[s[j]];
            ++j;
            while (cnt > 0) {
                --A[s[i]];
                if (A[s[i]] > 0) {
                    --cnt;
                }
                ++i;
            }
            if (result < j - i) {
                result = j - i;
            }
        }
        return result;
    }
    public static void Main() {
        string s;
        int result;

        s = "abcabcbb";
        result = LengthOfLongestSubstring(s);
        Console.WriteLine(result);

        s = "bbbbb";
        result = LengthOfLongestSubstring(s);
        Console.WriteLine(result);

        s = "pwwkew";
        result = LengthOfLongestSubstring(s);
        Console.WriteLine(result);
    }
}
